File: Medium/003-longest-substring-without-repeating-characters/python/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # Implement jump version
        char_index_map = {} # stores { character: index }
        left = 0
        max_len = 0

        for right, char in enumerate(s):
            if char in char_index_map:
                # Jump 'left' to the right of the previous occurrence.
                # max() ensures we never jump BACKWARDS if the duplicate is outside the window.
                left = max(char_index_map[char] + 1, left)
            
            # Update the map with the most recent position of this character
            char_index_map[char] = right

            max_len = max(max_len, right - left + 1)

        return max_len
            

File: Medium/003-longest-substring-without-repeating-characters/javascript/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    // Implement jump version using a set in js
    let charIndexSet = new Map()
    let left = 0
    let maxLen = 0

    for(var right = 0; right < s.length; right++) {
        const char = s[right];
        if(charIndexSet.has(char)) {
            // Jump 'left' to the right of the previous occurrence.
            // max() ensures we never jump BACKWARDS if the duplicate is outside the window.
            left = charIndexSet.get(char) + 1 > left ? charIndexSet.get(char) + 1 : left
            
        }
        // Update the map with the most recent position of this character
        charIndexSet.set(char, right)

        maxLen = maxLen > right - left + 1 ? maxLen : right - left + 1

    }
    return maxLen
};

module.exports = lengthOfLongestSubstring